financial-templates-lib/ExpiringMultiPartyEventClient.js

```diff
@@ -0,0 +1,168 @@
+
+// A thick client for getting information about an ExpiringMultiParty events.
+// This client is kept separate from the main ExpiringMultiPartyClient to keep
+// a clear separation of concerns and to limit the overhead from querying chain necessarily.
+// If no updateThreshold is specified then default to updating every 60 seconds.
+class ExpiringMultiPartyEventClient {
+  constructor(abi, web3, empAddress, updateThreshold = 60) {
+    this.updateThreshold = updateThreshold;
+    this.lastUpdateTimestamp;
+
+    this.web3 = web3;
+
+    // EMP contract
+    this.emp = new web3.eth.Contract(abi, empAddress);
+    this.empAddress = empAddress;
+
+    // EMP Events
+    this.liquidationEvents = [];
+    this.disputeEvents = [];
+    this.disputeSettlementEvents = [];
+  }
+
+  // Calls _update unless it was recently called, as determined by this.updateThreshold.
+  update = async () => {
+    const currentTime = Math.floor(Date.now() / 1000);
+    if (currentTime < this.lastUpdateTimestamp + this.updateThreshold) {
+      Logger.debug({
+        at: "ExpiringMultiPartyEventClient",
+        message: "EMP state update skipped",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTimestamp,
+        timeRemainingUntilUpdate: this.lastUpdateTimestamp + this.updateThreshold - currentTime
+      });
+      return;
+    } else {
+      await this._update();
+      this.lastUpdateTimestamp = currentTime;
+      Logger.debug({
+        at: "ExpiringMultiPartyEventClient",
+        message: "EMP state updated",
+        lastUpdateTimestamp: this.lastUpdateTimestamp
+      });
+    }
+  };
+
+  // Force call of _update, designed to be called by downstream caller that knowingly updated the EMP state.
+  forceUpdate = async () => {
+    const currentTime = Math.floor(Date.now() / 1000);
+    await this._update();
+    this.lastUpdateTimestamp = currentTime;
+    Logger.debug({
+      at: "ExpiringMultiPartyEventClient",
+      message: "EMP state force updated",
+      lastUpdateTimestamp: this.lastUpdateTimestamp
+    });
+  };
+
+  // Delete all events within the client
+  clearState = async () => {
+    this.liquidationEvents = [];
+    this.disputeEvents = [];
+    this.disputeSettlementEvents = [];
+  };
+
+  // Returns an array of liquidation events.
+  getAllLiquidationEvents = () => this.liquidationEvents;
+
+  // Returns an array of dispute events.
+  getAllDisputeEvents = () => this.disputeEvents;
+
+  // Returns an array of dispute events.
+  getAllDisputeSettlementEvents = () => this.disputeSettlementEvents;
+
+  start = () => {
+    this._poll();
+  };
+
+  _poll = async () => {
+    while (true) {
+      try {
+        await this._update();
+      } catch (error) {
+        Logger.error({
+          at: "ExpiringMultiPartyEventClient",
+          message: "client polling error",
+          error: error
+        });
+      }
+      await delay(Number(10_000));
+    }
+  };
```

Comment:
This is not strictly needed. It's here because I followed the pattern from the `ExpiringMultiPartyClient.js` which still has it [here](https://github.com/UMAprotocol/protocol/blob/d16d4f0ddc56864eb749ba90abb82fadbd28bd87/financial-templates-lib/ExpiringMultiPartyClient.js#L98). 

More broadly I plan on refactoring how these clients work in a later PR. Specifically The `ExpiringMultiPartyClient.js`, `ExpiringMultiPartyEventClient.js` and `GasEstimator.js` all use the same logic for polling the chain with a cooldown delay from it's previous poll in the `update` function. I want to refactor these to all extend a base class that implements this polling logic and then each client overrides the `_update()` function to implement it's specific clients data. I have an issue for this [here](https://github.com/UMAprotocol/protocol/issues/1196)

I included this code here for now to keep inline with convention of the other Client. I will remove it when I deal with that issue.

Reply:
That SGTM. I like the idea of creating a universal base class -- +1.